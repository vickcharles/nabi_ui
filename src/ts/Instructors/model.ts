export interface ExperienceType {
  id?: number;
  employer: string;
  jobTitle: string;
  fromMonth: string;
  fromYear: string;
  toMonth: string;
  toYear: string;
  location: string;
}

export enum DegreeType {
  associate = 'Associate Degree',
  bachelors = 'Bachelor\'s Degree',
  graduate = 'Graduate Degreee',
  professional = 'Professional Degree',
  certification = 'Certification',
  other = 'Other'
}

export interface EducationType {
  id?: number;
  schoolName: string;
  graduationYear: string;
  degreeType: DegreeType;
  fieldOfStudy: string;
}

export enum PlaceForLessonsOptions {
  home = 'Student\'s home',
  studio = 'Instructor\'s studio',
  online = 'Online',
}

export interface PlaceForLessonsType {
  home: boolean;
  studio: boolean;
  online: boolean;
}

export enum LessonDuration {
  thirtyMins = '30 mins',
  fortyFiveMins = '45 mins',
  sixtyMins = '60 mins',
  ninetyMins = '90 mins'
}

export interface RatesType {
  thirtyMinsRate: number;
  fortyFiveMinsRate: number;
  sixtyMinsRate: number;
  ninetyMinsRate: number;
}

export enum SkillLevel {
  beginner = 'Beginner',
  intermediate = 'Intermediate',
  advanced = 'Advanced'
}

export interface InstrumentsType {
  instrument: string;
  skillLevel: SkillLevel;
}

export interface InstructorType {
  userId: string;
  bio?: string;
  instruments?: InstrumentsType[];
  rates?: RatesType;
  placeForLessons?: PlaceForLessonsType;
  studioAddress?: string;
  education?: EducationType[];
  experience?: ExperienceType[];
}

export type IState = InstructorType[];
