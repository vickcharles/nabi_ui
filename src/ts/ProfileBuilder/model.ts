export interface InstructorState {
  userId: string;
  bio?: string;
}

export type IState = InstructorState[];
  