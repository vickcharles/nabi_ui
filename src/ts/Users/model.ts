export enum HearAboutUs {
  Google = 'google',
  Facebook = 'facebook',
  Craigslist = 'craigslist',
  PrintAds = 'printAds',
  WordOfMouth = 'wordOfMouth',
  CurrentNabiUser = 'currentNabiUser',
  Other= 'other'
}

export interface HearAboutUsInfo {
  value: HearAboutUs;
  label: string;
}

export const hearAboutUsInfo = {
  Google: <HearAboutUsInfo> {
    value: HearAboutUs.Google,
    label: 'Google'
  },
  Facebook: <HearAboutUsInfo> {
    value: HearAboutUs.Facebook,
    label: 'Facebook'
  },
  Craigslist: <HearAboutUsInfo> {
    value: HearAboutUs.Craigslist,
    label: 'Craigslist'
  },
  PrintAds: <HearAboutUsInfo> {
    value: HearAboutUs.PrintAds,
    label: 'Print Ads'
  },
  WordOfMouth: <HearAboutUsInfo> {
    value: HearAboutUs.WordOfMouth,
    label: 'Word Of Mouth'
  },
  CurrentNabiUser: <HearAboutUsInfo> {
    value: HearAboutUs.CurrentNabiUser,
    label: 'Current Nabi User'
  },
  Other: <HearAboutUsInfo> {
    value: HearAboutUs.Other,
    label: 'Other'
  }
};

export enum Role {
  student = 'student',
  instructor = 'instructor',
  parent = 'parent'
}

export interface UserState {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  password: string;
  zipCode: string;
  role: Role;
  hearAboutUs: string;
  avatar?: string;
  displayName: string;
}

export type IState = UserState[];
