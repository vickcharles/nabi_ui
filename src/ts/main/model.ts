/**
 * State for firing redirect
 * @interface RedirectState
 */
export interface RedirectState {
  fireRedirect: boolean;
}
