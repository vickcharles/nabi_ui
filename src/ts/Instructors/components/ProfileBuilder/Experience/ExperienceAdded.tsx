import * as React from 'react';
import IconButton from 'material-ui/IconButton';
import DeleteIcon from '@material-ui/icons/Delete';
import Typography from 'material-ui/Typography';
import { ExperienceType }  from '../../../model';
import Divider from 'material-ui/Divider';

/**
 * Props for ExperienceAdded
 * @interface ExperienceAddedProps
 */
interface ExperienceAddedProps extends ExperienceType {
  deleteExperience: (experienceId: any) => void;
}

/**
 * ExperienceAdded
 * @extends React.StatelessComponent<ExperienceAddedProps>
 */
const ExperienceAdded: React.StatelessComponent<ExperienceAddedProps> = props => {
  const fromMonth = props.fromMonth.charAt(0).toUpperCase() + props.fromMonth.slice(1);
  const toMonth = props.toMonth.charAt(0).toUpperCase() + props.toMonth.slice(1);
  return (
    <div>
      <div className="selected-education">
        <IconButton className="nabi-float-right" aria-label="Delete" onClick={() => props.deleteExperience(props.id)}>
          <DeleteIcon />
        </IconButton>
        <Typography className="nabi-margin-top-small" variant="body2">{props.jobTitle.toUpperCase()}</Typography>
        <Typography className="">{props.employer}</Typography>
        <Typography className="">{`${fromMonth}, ${props.fromYear} - ${toMonth}, ${props.toYear}`}</Typography>
        <Typography className="">
          {props.location}
        </Typography>
      </div>
      <Divider className="nabi-margin-top-small"/>
    </div>
  );
};

export default ExperienceAdded;
