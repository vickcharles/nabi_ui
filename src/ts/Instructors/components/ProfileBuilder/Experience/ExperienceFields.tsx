import * as React from 'react';

import TextField from 'material-ui/TextField';
import { FormControl } from 'material-ui/Form';
import Input, { InputLabel } from 'material-ui/Input';
import Select from 'material-ui/Select';
import { MenuItem } from 'material-ui/Menu';
import Button from 'material-ui/Button';
import Divider from 'material-ui/Divider';

import { ExperienceType } from '../../../model';
import { months } from '../../../../../assets/data/months';

/**
 * Props for ExperienceFields
 * @interface ExperienceFieldsProps
 */
interface ExperienceFieldsProps extends ExperienceType {
  handleChange: (event: React.FormEvent<{}>) => void;
  addExperience: (event: React.FormEvent<{}>) => void;
}

/**
 * ExperienceFields
 * @extends React.StatelessComponent<ExperienceFieldProps>
 */
const ExperienceFields: React.StatelessComponent<ExperienceFieldsProps> = props => {
  const monthSelectItems = months.map(month => {
    const monthName = month.charAt(0).toUpperCase() + month.slice(1);
    return (
      <MenuItem key={month} value={month}>{monthName}</MenuItem>
    );
  });
  return (
    <div>
      <div className="nabi-experience">
        <TextField
          onChange={props.handleChange}
          fullWidth={true}
          id="employer"
          margin="normal"
          name="employer"
          placeholder="Company/ employer name"
          required={true}
          value={props.employer}
        />
        <TextField
          onChange={props.handleChange}
          className=""
          fullWidth={true}
          id="jobTitle"
          margin="normal"
          name="jobTitle"
          placeholder="Job title"
          required={true}
          value={props.jobTitle}
        />
        <FormControl className="nabi-instruments-select">
          <InputLabel htmlFor="fromMonth">From</InputLabel>
          <Select
            input={<Input id="fromMonth" name="fromMonth" />}
            value={props.fromMonth}
            onChange={props.handleChange}
          >
            {monthSelectItems}
          </Select>
        </FormControl>
        <TextField
          onChange={props.handleChange}
          id="fromYear"
          margin="normal"
          name="fromYear"
          placeholder="Year"
          required={true}
          value={props.fromYear}
        />
        <FormControl className="nabi-instruments-select">
          <InputLabel htmlFor="toMonth">To</InputLabel>
          <Select
            input={<Input id="toMonth" name="toMonth" />}
            value={props.toMonth}
            onChange={props.handleChange}
          >
            {monthSelectItems}
          </Select>
        </FormControl>
        <TextField
          onChange={props.handleChange}
          id="toYear"
          margin="normal"
          name="toYear"
          placeholder="Year"
          required={true}
          value={props.toYear}
        />
        <TextField
          onChange={props.handleChange}
          fullWidth={true}
          id="location"
          margin="normal"
          name="location"
          placeholder="Job location"
          required={true}
          value={props.location}
        />
        <Button 
          color="primary"
          variant="raised"
          className="nabi-display-block nabi-margin-top-small"
          onClick={props.addExperience}
        >
          Add
        </Button>
      </div>
      <Divider className="nabi-margin-top-small"/>
    </div>
  );
};

export default ExperienceFields;
