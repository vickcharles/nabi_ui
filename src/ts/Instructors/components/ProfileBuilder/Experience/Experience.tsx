import * as React from 'react';
import { Action, Dispatch } from 'redux';
import { connect } from 'react-redux';

import Typography from 'material-ui/Typography';

import ExperienceFields from './ExperienceFields';
import { ExperienceType, InstructorType } from '../../../model';
import { updateInstructor } from '../../../';
import ExperienceAdded from './ExperienceAdded';

/**
 * Experience dispatch props
 * @interface ExperienceDispatchProps
 */
interface ExperienceDispatchProps {
  updateInstructor: (instructor: InstructorType) => void;
}

/**
 * Experience own props
 * @interface ExperienceOwnProps
 */
interface ExperienceOwnProps {
  userId: string;
}

/**
 * Experience props
 * @interface ExperienceProps
 * @extends BasicInfoDispatchProps & BasicInfoOwnProps
 */
interface ExperienceProps extends
  ExperienceDispatchProps,
  ExperienceOwnProps { }

/**
 * State for Experience
 * @interface ExperienceState
 * @extends ExperienceType
 */
interface ExperienceState extends ExperienceType {
  experience: ExperienceType[];
}

/**
 * Experience: Third step of the Profile Builder
 * @class Experience
 * @extends React.Component<ExperienceProps, ExperienceState>
 */
export class Experience extends React.Component<ExperienceProps, ExperienceState > {
  constructor(props: ExperienceProps) {
    super(props);

    this.state = {
      id: 0,
      employer: '',
      jobTitle: '',
      fromMonth: 'jan',
      fromYear: '',
      toMonth: 'jan',
      toYear: '',
      location: '',
      experience: [],
    };

    this.handleChange = this.handleChange.bind(this);
    this.addExperience = this.addExperience.bind(this);
  }

  public handleChange(event: any): void {
    const target = event.target;
    const value = target.value;
    const name = target.name;

    this.setState({
      ...this.state,
      [name]: value
    });
  }

  public addExperience(): void {
    const experienceToAdd: ExperienceType = {
      id: this.state.experience.length,
      employer: this.state.employer,
      jobTitle: this.state.jobTitle,
      fromMonth: this.state.fromMonth,
      fromYear: this.state.fromYear,
      toMonth: this.state.toMonth,
      toYear: this.state.toYear,
      location: this.state.location
    };

    if (this.state.employer && this.state.jobTitle && this.state.fromMonth && this.state.fromYear
      && this.state.toMonth && this.state.toYear && this.state.location
    ) {
      this.setState(
        { experience: [...this.state.experience, experienceToAdd] },
        () => {
        const instructor: InstructorType =  {
          userId: this.props.userId,
          experience: this.state.experience
        };
        this.updateInstructorCall(instructor);
        this.clearExperience();
      });
    }
  }

  public updateInstructorCall(instructor: InstructorType): void {
    this.props.updateInstructor(instructor);
  }

  public clearExperience(): void {
    this.setState({  
      employer: '',
      jobTitle: '',
      fromMonth: 'jan',
      fromYear: '',
      toMonth: 'jan',
      toYear: '',
      location: ''
    });
  }

  public deleteExperience(experienceId: number): void {
    this.setState(
      { experience: this.state.experience.filter(
        experience => experience.id !== experienceId)},
      () => {
      const instructor: InstructorType =  {
        userId: this.props.userId,
        experience: this.state.experience
      };
      this.updateInstructorCall(instructor);
    });
  }

  public render(): JSX.Element {
    const experienceAdded = this.state.experience.map((experience, i) => (
      <ExperienceAdded
        key={i}
        id={experience.id}
        employer={experience.employer}
        jobTitle={experience.jobTitle}
        fromMonth={experience.fromMonth}
        fromYear={experience.fromYear}
        toMonth={experience.toMonth}
        toYear={experience.toYear}
        location={experience.location}
        deleteExperience={(experienceId: number) => this.deleteExperience(experienceId)}
      />
    ));
    return (
      <div>
        {experienceAdded}
        <Typography className="nabi-margin-top-small" variant="body2">
          Add Experience
        </Typography>
        <ExperienceFields 
          handleChange={this.handleChange} 
          addExperience={this.addExperience} 
          employer={this.state.employer}
          jobTitle={this.state.jobTitle}
          fromMonth={this.state.fromMonth}
          fromYear={this.state.fromYear}
          toMonth={this.state.toMonth}
          toYear={this.state.toYear}
          location={this.state.location}
        />
      </div>
    );
  }
}

function mapDispatchToProps(
  dispatch: Dispatch<Action>,
  _ownProps: ExperienceOwnProps
): ExperienceDispatchProps {
  return {
    updateInstructor: (instructor: InstructorType) => dispatch(updateInstructor(instructor)),
  };
}

export default connect(null, mapDispatchToProps)(Experience);
