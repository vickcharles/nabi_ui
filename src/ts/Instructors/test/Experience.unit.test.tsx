import * as React from 'react';
import { Experience } from '../components/ProfileBuilder/Experience/Experience';
import { shallow } from 'enzyme';
import { InstructorType } from '../model';

describe('Experience', () => {
  let wrapper: any;
  const updateInstructor: (instructor: InstructorType) => {} = jest.fn();
  const updateInstructorCall = jest.spyOn(Experience.prototype, 'updateInstructorCall');
  const clearExperience = jest.spyOn(Experience.prototype, 'clearExperience');

  describe('Shallow', () => {
    beforeEach(() => {
      wrapper = shallow(<Experience userId="" updateInstructor={updateInstructor} />);
    });
    
    it('Matches snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Method handleChange()', () => {
    function test_handleChange(theName: string, theValue: any): void {
      describe(`When the event contains ${theName} for name and ${theValue} for value`, () => {
        beforeEach(() => {
          const e = {
            target: {
              name: theName,
              value: theValue
            }
          };

          wrapper.instance().handleChange(e);
        });

        it(`Sets the state's ${theName} to ${theValue}`, () => {
          expect(wrapper.state(theName)).toBe(theValue);
        });
      });
    }

    test_handleChange('employer', 'Oracle');
    test_handleChange('jobTitle', 'Software Engineer');
  });

  describe('Method addExperience()', () => {
    describe('When id, employer, jobTitle, fromMonth, fromYear, toMonth, toYear and location state are missing', () => {
      beforeEach(() => {
        wrapper.instance().addExperience();
      });

      it('Does not add experience()', () => {
        expect(wrapper.state('experience').length).toBe(0);
      });
      
      it('Does not call updateInstructorCall', () => {
        expect(updateInstructorCall).not.toBeCalled();
      });
    });

    describe('When id, employer, jobTitle, fromMonth, fromYear, toMonth, toYear and location state are defined', () => {
      beforeEach(() => {
        wrapper.setState({
          employer: 'Oracle',
          jobTitle: 'Software Engineer',
          fromMonth: 'sep',
          fromYear: '2017',
          toMonth: 'sep',
          toYear: '2018',
          location: 'Cambridge, MA'
        });
        wrapper.instance().addExperience();
      });

      it('Adds experience()', () => {
        expect(wrapper.state('experience').length).toBe(1);
      });
      
      it('Calls updateInstructorCall', () => {
        expect(updateInstructorCall).toBeCalled();
      });

      it('Calls clearExperience', () => {
        expect(clearExperience).toBeCalled();
      });
    });
  });

  describe('Method clearExperience()', () => {
    beforeEach(() => {
      wrapper.instance().clearExperience();
    });

    it('Clears experience', () => {
      expect(wrapper.state('employer')).toBe('');
      expect(wrapper.state('jobTitle')).toBe('');
      expect(wrapper.state('fromMonth')).toBe('jan');
      expect(wrapper.state('fromYear')).toBe('');
      expect(wrapper.state('toMonth')).toBe('jan');
      expect(wrapper.state('toYear')).toBe('');
      expect(wrapper.state('location')).toBe('');
    });
  });

  describe('Method deleteExperience()', () => {
    beforeEach(() => {
      wrapper.instance().deleteExperience();
    });

    it('Calls updateInstructorCall', () => {
      expect(updateInstructorCall).toBeCalled();
    });
  });
});
