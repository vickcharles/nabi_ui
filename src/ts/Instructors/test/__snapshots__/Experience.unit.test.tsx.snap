// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Experience Shallow Matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Experience
    updateInstructor={[MockFunction]}
    userId=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [],
        <WithStyles(Typography)
          className="nabi-margin-top-small"
          variant="body2"
        >
          Add Experience
        </WithStyles(Typography)>,
        <ExperienceFields
          addExperience={[Function]}
          employer=""
          fromMonth="jan"
          fromYear=""
          handleChange={[Function]}
          jobTitle=""
          location=""
          toMonth="jan"
          toYear=""
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Add Experience",
          "className": "nabi-margin-top-small",
          "variant": "body2",
        },
        "ref": null,
        "rendered": "Add Experience",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "addExperience": [Function],
          "employer": "",
          "fromMonth": "jan",
          "fromYear": "",
          "handleChange": [Function],
          "jobTitle": "",
          "location": "",
          "toMonth": "jan",
          "toYear": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [],
          <WithStyles(Typography)
            className="nabi-margin-top-small"
            variant="body2"
          >
            Add Experience
          </WithStyles(Typography)>,
          <ExperienceFields
            addExperience={[Function]}
            employer=""
            fromMonth="jan"
            fromYear=""
            handleChange={[Function]}
            jobTitle=""
            location=""
            toMonth="jan"
            toYear=""
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Add Experience",
            "className": "nabi-margin-top-small",
            "variant": "body2",
          },
          "ref": null,
          "rendered": "Add Experience",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "addExperience": [Function],
            "employer": "",
            "fromMonth": "jan",
            "fromYear": "",
            "handleChange": [Function],
            "jobTitle": "",
            "location": "",
            "toMonth": "jan",
            "toYear": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
